File: Constrain_Device/python-components/src/main/python/programmingtheiot/cda/system/SystemPerformanceManager.py
#####
# 
# This class is part of the Programming the Internet of Things project.
# 
# It is provided as a simple shell to guide the student and assist with
# implementation for the Programming the Internet of Things exercises,
# and designed to be modified by the student as needed.
#

import logging

from apscheduler.schedulers.background import BackgroundScheduler

import programmingtheiot.common.ConfigConst as ConfigConst

from programmingtheiot.common.ConfigUtil import ConfigUtil
from programmingtheiot.common.IDataMessageListener import IDataMessageListener

from programmingtheiot.cda.system.SystemCpuUtilTask import SystemCpuUtilTask
from programmingtheiot.cda.system.SystemMemUtilTask import SystemMemUtilTask

from programmingtheiot.data.SystemPerformanceData import SystemPerformanceData

class SystemPerformanceManager(object):
	"""
	This class is responsible for managing all the system performance related data. It primarly manager
	
	1) CpuUtilization
	2) MemoryUtilization

	It uses startManager and stopManager methods to run System Performance management in background

	It sets all the constants for system data and starts a background scheduler adding instances of SystemCpuUtilTask and SystemMemUtilTask

	It gets the telemetry values using handle telemetry method which invokes getTelemetryValues from the repective classes 
	
	"""

	def __init__(self):
		#pass
		configUtil = ConfigUtil()

		"""Setting All the class scope variables \
			-> pollRate
			-> locationID
		"""
		self.pollRate = \
		configUtil.getInteger(section = ConfigConst.CONSTRAINED_DEVICE, key = ConfigConst.POLL_CYCLES_KEY, defaultVal = ConfigConst.DEFAULT_POLL_CYCLES)
	
		self.locationID = \
		configUtil.getProperty(section = ConfigConst.CONSTRAINED_DEVICE, key = ConfigConst.DEVICE_LOCATION_ID_KEY, defaultVal = ConfigConst.NOT_SET)
	
		if self.pollRate <= 0:
			self.pollRate = ConfigConst.DEFAULT_POLL_CYCLES
		
		self.dataMsgListener = None

		'''*****Connecting SystemCpuUtilTask and SystemMemUtilTask to SystemPerformanceManager*****'''
		
		self.scheduler = BackgroundScheduler()
		self.scheduler.add_job(self.handleTelemetry, 'interval', seconds = self.pollRate)

		self.cpuUtilTask = SystemCpuUtilTask()
		self.memUtilTask = SystemMemUtilTask()


	def handleTelemetry(self):
		# pass

		'''Getting CPU and MEMORY Telemetry values'''
		self.cpuUtilPct = self.cpuUtilTask.getTelemetryValue()
		self.memUtilPct = self.memUtilTask.getTelemetryValue()

		logging.debug('\nCPU utilization is %s percentage\nMemory utilization is %s percentage',str(self.cpuUtilPct),str(self.memUtilPct))

		sysPrefData = SystemPerformanceData()
		sysPrefData.setLocationID(self.locationID)
		sysPrefData.setCpuUtilization(self.cpuUtilPct)
		sysPrefData.setMemoryUtilization(self.memUtilPct)

		if self.dataMsgListener:
			self.dataMsgListener.handleSystemPerformanceMessage(data = sysPrefData)

		
	def setDataMessageListener(self, listener: IDataMessageListener) -> bool:
		if listener:
			self.dataMsgListener = listener
	
	def startManager(self):
		#pass
		logging.info("Starting SystemPerformanceManager...")
		# Checking if scheduler is running 
		if not self.scheduler.running:
			self.scheduler.start()
			logging.info("Started SystemPerformanceManager.")
		else:
			logging.warning("SystemPerformanceManager scheduler already started. Ignoring....")
		
	def stopManager(self):
		#pass
		logging.info("Stopping SystemPerformanceManager...")
		
		try:
			self.scheduler.shutdown()
			logging.info("Stopped SystemPerformanceManager.")
		except:
			logging.warning("SystemPerformanceManager schedular already stopped. Ignoring...")

File: Gateway_Device/java-components/src/main/java/programmingtheiot/gda/system/SystemPerformanceManager.java
/**
 * This class is part of the Programming the Internet of Things project.
 * 
 * It is provided as a simple shell to guide the student and assist with
 * implementation for the Programming the Internet of Things exercises,
 * and designed to be modified by the student as needed.
 */ 

package programmingtheiot.gda.system;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ConfigUtil;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.gda.connection.IPersistenceClient;
import programmingtheiot.gda.connection.IPersistenceListener;
import programmingtheiot.gda.connection.RedisPersistenceAdapter;
import programmingtheiot.common.ResourceNameEnum;
import programmingtheiot.data.SystemPerformanceData;


/* Importing Logging framework */
import java.util.logging.Logger;
import java.util.logging.Level;

/**
 * This class creates the instances of all system performance classes.
 * It sets all the required paramaters and calls respective get telemetry methods
 * 
 */
public class SystemPerformanceManager
{
	// private var's
	/* Adding private class scope variables */
	private static final Logger _Logger = Logger.getLogger(SystemPerformanceManager.class.getName());
	private int pollRate = ConfigConst.DEFAULT_POLL_CYCLES;

	private ScheduledExecutorService schedExeSvc = null;
	private SystemCpuUtilTask sysCpuUtilTask = null;
	private SystemMemUtilTask sysMemUtilTask = null;
	private SystemDiskUtilTask sysDiskUtilTask = null;

	private Runnable taskRunner = null;
	private boolean isStarted = false;

	private String locationID = ConfigConst.NOT_SET;
	private IDataMessageListener dataMsgListener = null;
	private IPersistenceClient redIPersistenceClient = null;

	
	// constructors
	
	/**
	 * Default.
	 * 
	 */
	public SystemPerformanceManager()
	{
		this.pollRate = ConfigUtil.getInstance().getInteger(ConfigConst.GATEWAY_DEVICE, ConfigConst.POLL_CYCLES_KEY, ConfigConst.DEFAULT_POLL_CYCLES);
		if (this.pollRate <=0){
			this.pollRate = ConfigConst.DEFAULT_POLL_CYCLES;
		}

		this.locationID = ConfigUtil.getInstance().getProperty(ConfigConst.GATEWAY_DEVICE, ConfigConst.LOCATION_ID_PROP, ConfigConst.NOT_SET);

		this.schedExeSvc = Executors.newScheduledThreadPool(1);

		this.sysCpuUtilTask = new SystemCpuUtilTask();
		this.sysMemUtilTask = new SystemMemUtilTask();
		this.sysDiskUtilTask = new SystemDiskUtilTask();

		this.taskRunner = ()->{
			this.handleTelemetry();
		};
	}
	
	
	// public methods

	/* Getting CPU utilization and Memory Utilization values from classes
	 * -> SystemCpuUtilTask()
	 * -> SystemMemoryUtilTask()
	 * -> SystemDiskUtilTask()
	 */
	
	public void handleTelemetry()
	{
		float cpuUtil = this.sysCpuUtilTask.getTelemetryValue();
		float memUtil = this.sysMemUtilTask.getTelemetryValue();
		float diskUtil = this.sysDiskUtilTask.getTelemetryValue();

		RedisPersistenceAdapter redisClient = new RedisPersistenceAdapter();


		_Logger.info("CPU utiliztion: "+cpuUtil + "\nMemory Utilization: " + memUtil +"\nDisk Utilization: "+diskUtil);

		SystemPerformanceData spd = new SystemPerformanceData();

		spd.setLocationID(this.locationID);
		spd.setCpuUtilization(cpuUtil);
		spd.setMemoryUtilization(memUtil);
		spd.setDiskUtilization(diskUtil);

		redisClient.storeData(ResourceNameEnum.GDA_SYSTEM_PERF_MSG_RESOURCE.getResourceName(), ConfigConst.DEFAULT_QOS, spd);
		
		if (this.dataMsgListener != null)
		{
			this.dataMsgListener.handleSystemPerformanceMessage(ResourceNameEnum.GDA_SYSTEM_PERF_MSG_RESOURCE, spd);

		}
	}
	
	public void setDataMessageListener(IDataMessageListener listener)
	{
		if (listener != null)
		{
			this.dataMsgListener = listener;
		}
	}


/* Adding Start manager and stop manager for logging info */
	public boolean startManager()
	{
		if(! this.isStarted){
			_Logger.info("SystemPerformanceManager is Starting...");

			ScheduledFuture<?> futureTask = this.schedExeSvc.scheduleAtFixedRate(this.taskRunner, 1L, this.pollRate, TimeUnit.SECONDS);
			
			this.isStarted = true;
		}
		else{
			_Logger.info("SystemPerformanceManager is already started.");
		}

		return this.isStarted;
	}

	public boolean stopManager()
	{
		this.schedExeSvc.shutdown();
		this.isStarted = false;

		_Logger.info("SystemPerformanceManager is stopped.");

		return true;
	}
	
}
